README.md
# Utils

[![CodeQL](https://github.com/JasonMahdjoub/Utils/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/JasonMahdjoub/Utils/actions/workflows/codeql-analysis.yml)

Set of Java tools :
* Decentralized identifier generation (similar to UUID, but with hash possibilities and some enforcement)
* XML, YAML and properties class mapping
* Timer
* Get size of Java memory objects
* OS tools (OS version, trace route, hardrive tools, network speed)
* DNS checking and email validation tool
* Progress monitor with progress bar wrapper to enable multi-plateform compatibility
* Cryptography tools
	* Home interface that use of three Implementations : Java native JCE, GNU CRYPT, BouncyCastle (+ BouncyCastle FIPS and BouncyCastle PQC)
	* Encryption algorithms does not need signed JAR to work. So the release work on official Oracle JVM.
	* MessageDigest algorithms
	* Symmetric signatures 
	* Asymmetric signatures
	* Symmetric encryptions
	* Asymmetric encryptions and key wrappers
	* Key agreements
	* DRBG
	* JPAKE algorithm and home secret message exchanger
	* Login thanks to a symmetric signature process
	* Login thanks to an asymmetric signature process
	* Use of post quantum algorithms (New Hope key exchanger, Sphincs, McEliece)
	* Availability of hybrid assymetric key pairs. One hybrid key pair assemble two key pairs : one used with a non post quantum algorithm, and one another used with a post quantum algorithm. Hybrid algorithms like asymmetric encrytion, asymmetric signatures, key agreements, key wrappers, ... enable to use two algorithms (non PQC and PQC) at the same time, in order to keep well-tested non post quantum algorithms with new expérimental post quantum algorithms.
* Compatible with Java 8 and newer
* Random input/output streams (byte array, file). 
* Random cache file output stream which enables to write data into a byte array until an amount of data was reached. Then, when this amount was reached, use instead a file output stream.

# Changes

[See historical of changes](./changelog.md)

###### Requirements under Ubuntu/Debian :
  * Please install the package ethtool, rng-tools, mtr(only debian)

# How to use it ?

## With Gradle :

Adapt into your build.gradle file, the next code :
```
        ...
	dependencies {
		...
		api(group:'com.distrimind.util', name: 'Utils', version: '5.25.2-STABLE')
		...
		//choose one of these optional drivers for GnuCrypto algorithms
			testImplementation(group:'gnu', name: 'Gnu-Crypt', version: '1.3.0')

	}
	...
```
Librairies are available on Maven Central. You can check signatures of dependencies with this [public GPG key](./gpg_key.asc). You can also use the next repository : 
```
	...
	repositories {
		...
		maven {
	       		url "https://artifactory.distri-mind.fr:443/artifactory/gradle-release/"
	   	}
		...
	}
	...
```	
To know what last version has been uploaded, please refer to versions availables into [this repository](https://artifactory.distri-mind.fr/artifactory/DistriMind-Public/com/distrimind/util/Utils/)
## With Maven :
Adapt into your pom.xml file, the next code :
```
	...
	<project>
		...
		<dependencies>
			...
			<dependency>
				<groupId>com.distrimind.util</groupId>
				<artifactId>Utils</artifactId>
				<version>5.25.2-STABLE</version>
			</dependency>
			<!-- choose one of these optional drivers for GnuCrypto algorithms-->
			<dependency>
				<groupId>gnu</groupId>
				<artifactId>Gnu-Crypt</artifactId>
				<version>1.3.0</version>
			</dependency>
			...
		</dependencies>
		...

	</project>
	...
```	
Librairies are available on Maven Central. You can check signatures of dependencies with this [public GPG key](./gpg_key.asc). You can also use the next repository : 
```
	...
	<repositories>
		...
		<repository>
			<id>DistriMind-Public</id>
			<url>https://artifactory.distri-mind.fr:443/artifactory/gradle-release/</url>
		</repository>
		...
	</repositories>
	...		
```
To know what last version has been uploaded, please refer to versions availables into [this repository](https://artifactory.distri-mind.fr/artifactory/DistriMind-Public/com/distrimind/util/Utils/)





changelog.md
Utils
=====
5.25.2 STABLE (Build: 5041) (from 04/01/2016 to 10/01/2023)

# Creator(s):
* Jason MAHDJOUB

# Developer(s):
* Jason MAHDJOUB (Entered in the team at 04/01/2016)

# Changes:


### 5.25.2 STABLE (10/01/2023)
#### Internal change(s)
* Make WrappedSecretString.constantTimeAreEqual(...) function more generic, and use CharSequence values instead of String


### 5.25.1 STABLE (09/01/2023)
#### Internal change(s)
* Add transitives rules into module-info file


### 5.25.0 STABLE (06/01/2023)
#### New feature(s)
* Add WrappedHashedValue class
* Add WrappedHashedValueInBase64StringFormat class
* Add MessageDigestAlgorithmType enum
* Add Ed25519 and Ed448 java native implementations


### 5.24.2 STABLE (03/01/2023)
#### Internal change(s)
* Change language level to Java 11
* Update dependencies
* Clean code


### 5.24.1 STABLE (22/12/2022)
#### New feature(s)
* Add class SystemInfo
* Enable or disable side channel attacks accord to CPU type and permit to manually disable/enable mitigation through system properties
#### Internal change(s)
* Fix spelling errors
* Interface Cleanable do not inherit anymore the interface AutoCloseable
* MultiFormatProperties implements now SecureExternalizable instead of Serializable interface
#### Security fixe(s) with low severity
* Fix CodeQL error with RSA key size not checked.
#### Bug fixe(s)
* Fix issues with multi thread encryption and signature


### 5.24.0 STABLE (19/10/2022)
#### New feature(s)
* Add possibility to use several thread with SymmetricSignatureHashEncoder and SymmetricSignatureHashDecoder. Add functions SymmetricSignatureHashEncoder.withPoolExecutor(PoolExecutor) and SymmetricSignatureHashDecoder.withPoolExecutor(PoolExecutor).
* Add possibility to use encoder and decoder as continuous network stream. Main secret key is regenerated after each encoding.
* Add BouncyCastle Sphincs+ asymmetric authenticated signature post quantum algorithms.
* Add Chrystals-Kyber asymmetric key wrappers.
* Add NTRU asymmetric key wrappers.
* Add SABER asymmetric key wrappers.
* Add functions ASymmetricKeyWrapperType.getKeyPairGenerator(...).
* Add BouncyCastle Chrystals-Dilithium asymmetric authenticated signature post quantum algorithms.
* Add BouncyCastle Falcon asymmetric authenticated signature post quantum algorithms.
* Add key agreement that use key wrapping with class KeyAgreementWithKeyWrapping.
* Reorganise architecture of key agreements, and hybrid key agreements. Permit key agreement that generates both secret key for encryption and secret key for signature.
#### Internal change(s)
* Update BouncyCastle to 1.71
* Update BouncyCastle-FIPS to 1.0.2.3
* Clean code
* Improve nonce generator from seed into SecureRandomType
* Target java compatibility is set to Java 11 but source code still use Java 8
* Encode key types and algorithm types with their derived final type
#### Security fixe(s) with low severity
* Update snakeyaml dependency, and fix CVE-2022-38752, CVE-2022-38751, CVE-2022-38750, CVE-2022-38749
* Deprecate GNU algorithms because GNU dependency is no more updated into Utils.
#### Security fixe(s) with medium severity
* Add additional CPU usage during encryption using specifics algorithms to limit frequency side channel attacks and power side channel power attacks. Concerned encryption algorithm are referenced into class SymmetricEncryptionType. Asymmetric encryption are also concerned. Encryption that were used to encode little blocks are less concerned by these attacks.
* Generate new keys during encryption after using the same key with a predefined quantity of data to limit frequency side channel attacks and power side channel power attacks. Concerned encryption algorithm are referenced into class SymmetricEncryptionType. Asymmetric encryption are also concerned. Encryption that were used to encode little blocks are less concerned by these attacks.


### 5.23.5 STABLE (07/04/2022)
#### Internal change(s)
* Alter Cleanable API.
#### Bug fixe(s)
* Fix NullPointerException into finalizer of NewHopeKeyAgreementServer class.
* Fix issue with Cleaner on JVM 8. Finalize was not called. JVM greater than Java 8 was not concerned by this issue.
* Fix dead lock with Cleaner.


### 5.23.2 STABLE (05/04/2022)
#### Internal change(s)
* Add lacking files into generated libraries.
* Add constructors into classes Description and Version.


### 5.23.1 STABLE (05/04/2022)
#### Internal change(s)
* Update README files.


### 5.23.0 STABLE (04/04/2022)
#### New feature(s)
* Add interface Cleanable that use Cleanable java API when JVM version is greater than Java 8 and that use standard finalize method otherwise.
#### Internal change(s)
* Remove finalize method into all key classes, and make them using Cleanable API
#### Security fixe(s)
* Possible bad use of keys, if theses keys have been zeroized. Now an exception is thrown if these keys are used after being zeroized 


### 5.22.4 STABLE (29/03/2022)
#### Internal change(s)
* Declare dependencies as implementation instead of API gradle mode.


### 5.22.3 STABLE (25/03/2022)
#### Internal change(s)
* Improve Utils class loader
* Make MultiFormatProperties use UTF-8 as default encoding
* Zeroizable interface now extends Destroyable interface
#### Bug fixe(s)
* Fix compatibility problems with Android
* Fix array index error into IDGeneratorInt


### 5.22.2 STABLE (07/02/2022)
#### New feature(s)
* Fix bad android os detection with a false positive in some cases.
#### Internal change(s)
* Update URLs.
* Update Snakeyaml to 1.30
* Add detection of Android API 30 and 31


### 5.22.1 STABLE (25/01/2022)
#### Internal change(s)
* Do not use BouncyCastle RSA implementation with hybrid asymmetric key wrapper type.


### 5.22.0 STABLE (24/01/2022)
#### New feature(s)
* Add function MessageDigestType.isSecuredForSignature().
* Add hybrid asymmetric key wrappers.
* Add max variables into ASymmetricKeyWrapperType and into SymmetricKeyWrapperType class that gives max size in byte of wrapped keys.

src/com/distrimind/util/Utils.java
/*
Copyright or © or Copr. Jason Mahdjoub (04/02/2016)

jason.mahdjoub@distri-mind.fr

This software (Utils) is a computer program whose purpose is to give several kind of tools for developers 
(ciphers, XML readers, decentralized id generators, etc.).

This software is governed by the CeCILL-C license under French law and
abiding by the rules of distribution of free software.  You can  use, 
modify and/ or redistribute the software under the terms of the CeCILL-C
license as circulated by CEA, CNRS and INRIA at the following URL
"http://www.cecill.info". 

As a counterpart to the access to the source code and  rights to copy,
modify and redistribute granted by the license, users are provided only
with a limited warranty  and the software's author,  the holder of the
economic rights,  and the successive licensors  have only  limited
liability. 

In this respect, the user's attention is drawn to the risks associated
with loading,  using,  modifying and/or developing or reproducing the
software by the user in light of its specific status of free software,
that may mean  that it is complicated to manipulate,  and  that  also
therefore means  that it is reserved for developers  and  experienced
professionals having in-depth computer knowledge. Users are therefore
encouraged to load and test the software's suitability as regards their
requirements in conditions enabling the security of their systems and/or 
data to be ensured and,  more generally, to use and operate it in the 
same conditions as regards security. 

The fact that you are presently reading this means that you have had
knowledge of the CeCILL-C license and that you accept its terms.
 */
package com.distrimind.util;

import java.io.InputStream;
import java.util.Calendar;

import com.distrimind.util.export.License;
import com.distrimind.util.version.Description;
import com.distrimind.util.version.Person;
import com.distrimind.util.version.PersonDeveloper;
import com.distrimind.util.version.Version;

/**
 * 
 * @author Jason Mahdjoub
 * @version 1.7
 */
public class Utils
{
    public static final Version VERSION;

    public static final License LICENSE = new License(
	    License.PredefinedLicense.CeCILL_C_v1_0);
    static
    {
	Calendar c = Calendar.getInstance();
	c.set(2016, 1, 4);
	Calendar c2 = Calendar.getInstance();
	c.set(2017, 5, 19);
	VERSION = new Version("Utils", "Utils", 2, 10, 0, Version.Type.Stable, 0,
		c.getTime(), c2.getTime());
	try
	{
	    

	    InputStream is = Utils.class.getResourceAsStream("build.txt");
	    VERSION.loadBuildNumber(is);

	    VERSION.addCreator(new Person("mahdjoub", "jason"));
	    c = Calendar.getInstance();
	    c.set(2016, 1, 4);
	    VERSION.addDeveloper(
		    new PersonDeveloper("mahdjoub", "jason", c.getTime()));

	    c = Calendar.getInstance();
	    c.set(2017, 5, 19);
	    Description d = new Description(2, 10, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Adding symmetric signture algorithms");
	    d.addItem("Altereging P2PJPAKESecretMessageExchanger class");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2017, 5, 18);
	    d = new Description(2, 9, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Adding Elliptic Curve Diffie-Hellman key exchange support");
	    d.addItem("Password Authenticated Key Exchange by Juggling (2008) algorithm");
	    d.addItem("Adding Bouncy Castle algorithms");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2017, 5, 1);
	    d = new Description(2, 8, 0, Version.Type.Stable, 0,c.getTime());
	    d.addItem("Managing enum type into XML properties");
	    d.addItem("XML properties are able to manage abstract sub XML properties");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2017, 4, 23);
	    d = new Description(2, 7, 1, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Altering ListClasses");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2017, 4, 3);
	    d = new Description(2, 7, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Adding primitive tab support for XML Properties");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2017, 3, 24);
	    d = new Description(2, 6, 1, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("JDK 7 compatible");
	    d.addItem("Correcting a bug with testReadWriteDataPackaged in CryptoTests");
	    VERSION.addDescription(d);
	    
	    d = new Description(2, 6, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Adding RegexTools class");
	    d.addItem("JDK 7 compatible");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2017, 2, 7);
	    d = new Description(2, 5, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Improving and renforcing P2PAsymmetricSecretMessageExchanger");
	    d.addItem("Additional manifest content possibility for projects export");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2017, 2, 4);
	    d = new Description(2, 4, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Debugging documentation export");
	    d.addItem("Updating common net to 3.6 version");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2017, 1, 7);
	    d = new Description(2, 3, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("AbstractXMLObjectParser is now serializable");
	    VERSION.addDescription(d);
	    
	    
	    c = Calendar.getInstance();
	    c.set(2017, 0, 5);
	    d = new Description(2, 2, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Updating IDGeneratorInt class and fix memory leak problem");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 11, 31);
	    d = new Description(2, 1, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Adding expiration time for public keys");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 11, 23);
	    d = new Description(2, 0, 1, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Changing gnu cryto packages");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 11, 17);
	    d = new Description(2, 0, 0, Version.Type.Stable, 0,
		    c.getTime());
	    d.addItem("Including Gnu Crypto Algorithms.");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 11, 6);
	    d = new Description(1, 9, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem(
		    "Correcting a bug with the use of IV parameter. Now, the IV parameter is generated for each encryption.");
	    d.addItem("Adding class SecureRandomType.");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 9, 13);
	    d = new Description(1, 8, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding password hash (PBKDF and bcrypt)");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 8, 15);
	    d = new Description(1, 7, 2, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correcting a bug for P2PASymmetricSecretMessageExchanger");
	    d.addItem("Adding toString and valueOf functions for crypto keys");
	    d.addItem("Possibility to put crypto keys in XMLProperties class");
	    d.addItem("Adding 'valueOf' for Decentralized IDs");
	    d.addItem("Decentralized IDs are exportable into XML Properties");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 7, 23);
	    d = new Description(1, 7, 1, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correcting a bug for loop back network interface speed");
	    d.addItem(
		    "Correcting a bug for P2PASymmetricSecretMessageExchanger");
	    d.addItem("Correcting a bug big data asymmetric encryption");
	    d.addItem("Adding symmetric et asymmetric keys encapsulation");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 6, 4);
	    d = new Description(1, 7, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem(
		    "Renaming class ASymmetricEncryptionAlgorithm to P2PASymmetricEncryptionAlgorithm");
	    d.addItem("Adding class SignatureCheckerAlgorithm");
	    d.addItem("Adding class SignerAlgorithm");
	    d.addItem("Adding class ClientASymmetricEncryptionAlgorithm");
	    d.addItem("Adding class ServerASymmetricEncryptionAlgorithm");
	    d.addItem("Updating to Common-Net 3.5");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 5, 10);
	    d = new Description(1, 6, 1, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correcting bug into XMLProperties class");
	    d.addItem("Adding tests for XMLProperties class");
	    d.addItem("Changing license to CECILL-C.");
	    d.addItem("Correcting bugs into DecentralizedIDGenerator classes");
	    d.addItem("Adding salt management into SecuredIDGenerator class");
	    d.addItem(
		    "Adding salt management into PeerToPeerASymetricSecretMessageExanger class");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 2, 15);
	    d = new Description(1, 6, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding unit tests possibility for project export tools");
	    d.addItem("Adding unit compilation for project export tools");
	    d.addItem("Adding new licences");
	    VERSION.addDescription(d);
	    
	    c = Calendar.getInstance();
	    c.set(2016, 2, 9);
	    d = new Description(1, 5, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem(
		    "Adding PeerToPeerASymmetricSecretMessageExchanger class");
	    d.addItem(
		    "Adding ObjectSizer class (determins sizeof each java object instance)");
	    d.addItem("Adding keys encoding");
	    d.addItem("Adding decentralized id encoding/decoding");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2016, 2, 1);
	    d = new Description(1, 4, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding encryption utilities");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2016, 1, 24);
	    d = new Description(1, 3, 1, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Set Bits static functions public");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2016, 1, 22);
	    d = new Description(1, 3, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding SecuredDecentralizedID class");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2016, 1, 15);
	    d = new Description(1, 2, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding function AbstractXMLObjectParser.isValid(Class)");
	    d.addItem(
		    "Correcting export bug : temporary files were not deleted.");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2016, 1, 14);
	    d = new Description(1, 1, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding some internal modifications to ReadWriteLocker");
	    VERSION.addDescription(d);

	    c = Calendar.getInstance();
	    c.set(2016, 1, 4);
	    d = new Description(1, 0, 0, Version.Type.Stable, 0, c.getTime());
	    d.addItem("Realeasing first version of Utils");
	    VERSION.addDescription(d);
	}
	catch (Exception e)
	{
	    e.printStackTrace();
	}
    }

}
